pinot-core/src/main/java/org/apache/pinot/core/query/scheduler/QueryScheduler.java

```diff
@@ -168,6 +169,19 @@ protected byte[] processQueryAndSerialize(ServerQueryRequest queryRequest, Execu
 
       byte[] responseBytes = serializeResponse(queryRequest, instanceResponse);
 
+      Map<String, String> queryOptions = queryRequest.getQueryContext().getQueryOptions();
+      Long maxResponseByteLength = QueryOptionsUtils.getMaxSerializedResponseLengthPerServer(queryOptions);
+      if (maxResponseByteLength != null && responseBytes.length > maxResponseByteLength) {
+        String errorMessage = String.format("Serialized response exceeds threshold for requestId %d from broker %s",
+            queryRequest.getRequestId(), queryRequest.getBrokerId());
+        LOGGER.error(errorMessage);
+        // TODO(Vivek): Add a metric indicating the number of such exceptions.
+        instanceResponse = new InstanceResponseBlock();
+        instanceResponse.addException(QueryException.getException(QueryException.INTERNAL_ERROR, errorMessage));
```

Comment:
I think we should add a specific QueryException for this case

Reply:
or use QUERY_CANCELLATION_ERROR for which we've set the error code to 503(Service Unavailable)

Reply:
Good idea, done.